src/compile/layer.ts

```diff
@@ -108,27 +107,37 @@ export class LayerModel extends Model {
   }
 
   public parseAxisAndHeader() {
-    const axisComponent = this.component.axes = {};
+    const axisComponent: AxesComponent = this.component.axes = {};
 
     for (const child of this.children) {
       child.parseAxisAndHeader();
-      keys(child.component.axes).forEach(channel => {
-        // TODO: read these from the resolve syntax
-        const axisResolve = 'shared';
-        const scaleResolve = 'shared';
-
-        if (scaleResolve === 'shared' && axisResolve === 'shared') {
-          // If shared/union axis (only possible if the scale is shared in the first place)
+      keys(child.component.axes).forEach((channel: SpatialScaleChannel) => {
+        if (this.resolve[channel].axis === 'shared') {
+          // If shared/union axis
 
           // Just use the first axes definition for each channel
           // TODO: what if the axes from different children are not compatible
           if (!axisComponent[channel]) {
             axisComponent[channel] = child.component.axes[channel];
-            delete child.component.axes[channel];
           }
         } else {
-          // Otherwise do nothing for independent axes
+          // If axes are independent
+          if (!axisComponent[channel]) {
+            // copy the first axis
+            axisComponent[channel] = child.component.axes[channel];
+          } else if (axisComponent[channel].axes.length === 1) {
+            // and put the second one on the right/top
+            axisComponent[channel].axes.push({
+              ...child.component.axes[channel].axes[0],
+              orient: channel === 'y' ? 'right' : 'top'
+            });
+            axisComponent[channel].gridAxes.push({
+              ...child.component.axes[channel].gridAxes[0]
+            });
+          }
+          // Ignore all other axes.
```

Comment:
Why ignore all other axes?  This seems wrong.  At least showing them with occlusion and let users set offset manually is better than not showing. 

A better logic is probably alternating between left/right (or top/bottom) add incrementally add offset so the axes do not overlap.  However, I don't care that much about multiple axes anyway. Simply showing them as overlapping seems sufficient.